### PR TITLE
Add support for indirect descriptors and use them in vhost_user_fs

### PR DESCRIPTION
**Pull request by slp**
This patch series enables vhost_user_fs to use EVENT_IDX and INDIRECT_DESC virtio features, making it compatible with QEMU.

### CODE REVIEW

File: vm-virtio/src/queue.rs [hunk 1]
```diff
@@ -155,12 +157,61 @@ impl<'a> DescriptorChain<'a> {
...
+        }
+
+        let desc_head = self.addr;
+        self.mem.checked_offset(desc_head, 16)?;
+
+        // These reads can't fail unless Guest memory is hopelessly broken.
+        let desc = match self.mem.read_obj::<Descriptor>(desc_head) {
+            Ok(ret) => ret,
+            Err(_) => {
+                // TODO log address
```
Comment (rbradford) [inline vm-virtio/src/queue.rs]: Fix the TODO or drop it?

File: src/bin/vhost_user_fs.rs [hunk 2]
```diff
@@ -169,7 +183,23 @@ impl<F: FileSystem + Send + Sync + 'static> VhostUserBackend for VhostUserFsBack
             REQ_QUEUE_EVENT => {
                 debug!("REQ_QUEUE_EVENT");
                 let mut vring = vrings[1].write().unwrap();
-                self.process_queue(&mut vring)?;
+                if self.event_idx {
+                    // vm-virtio's Queue implementation only checks avail_index
```
Comment (rbradford) [inline src/bin/vhost_user_fs.rs]: Should change vm-virtio?

File: vm-virtio/src/queue.rs (hunk 1 above)
Comment (slp) [reply vm-virtio/src/queue.rs]: That's a copy/paste from `checked_new`. If we want to get rid of this one, we should probably get rid of both. Perhaps on a separate patch.

File: src/bin/vhost_user_fs.rs (hunk 2 above)
Comment (slp) [reply src/bin/vhost_user_fs.rs]: Actually, vm-virtio's Queue behavior is quite reasonable when not using EVENT_IDX. I think it's better to leave up to the callers to decide whether they want to check avail_index again by generating a new Queue iterator, or just do it once.

**Comment by rbradford** [comment]
One of the integration tests is timing out. I will spin up a VM to try and figure it out for you.

**Comment by rbradford** [comment]
@slp test_virtio_fs_dax_off_w_vhost_user_fs_daemon is one of the tests that's hanging. The backend process has exited but not been reaped.

**Comment by slp** [comment]
@rbradford Thanks for looking into this. Commit "vhost_user_fs: Be honest about protocol supported features" was missing the MQ feature. It should be working now.

File: src/bin/vhost_user_fs.rs [hunk 3]
```diff
@@ -172,7 +183,23 @@ impl<F: FileSystem + Send + Sync + 'static> VhostUserBackend for VhostUserFsBack
...
+                    // once, so to properly support EVENT_IDX we need to keep
+                    // calling process_queue() until it stops finding new
+                    // requests on the queue.
+                    loop {
+                        vring
+                            .mut_queue()
+                            .update_avail_event(self.mem.as_ref().unwrap());
+                        if !self.process_queue(&mut vring).unwrap() {
```
Comment (sboeuf) [inline src/bin/vhost_user_fs.rs]: Could you please make sure you're not breaking the error propagation by not using `unwrap()` functions?

File: src/bin/vhost_user_fs.rs [hunk 4]
```diff
@@ -172,7 +183,23 @@ impl<F: FileSystem + Send + Sync + 'static> VhostUserBackend for VhostUserFsBack
...
+                            break;
+                        }
+                    }
+                } else {
+                    // Without EVENT_IDX, a single call is enough.
+                    self.process_queue(&mut vring).unwrap();
```
Comment (sboeuf) [inline src/bin/vhost_user_fs.rs]: Ditto

File: vhost_user_fs/src/descriptor_utils.rs [hunk 5]
```diff
@@ -191,7 +191,12 @@ impl<'a> Reader<'a> {
     /// Construct a new Reader wrapper over `desc_chain`.
     pub fn new(mem: &'a GuestMemoryMmap, desc_chain: DescriptorChain<'a>) -> Result<Reader<'a>> {
         let mut total_len: usize = 0;
-        let buffers = desc_chain
+        let chain = if desc_chain.is_indirect() {
+            desc_chain.new_from_indirect().unwrap()
```
[the remark this answers is not in the document]
Comment (slp) [reply vhost_user_fs/src/descriptor_utils.rs]: Yeah, that'd better. Fixed.

File: vhost_user_fs/src/descriptor_utils.rs [hunk 6]
```diff
@@ -191,7 +194,14 @@ impl<'a> Reader<'a> {
     /// Construct a new Reader wrapper over `desc_chain`.
     pub fn new(mem: &'a GuestMemoryMmap, desc_chain: DescriptorChain<'a>) -> Result<Reader<'a>> {
         let mut total_len: usize = 0;
-        let buffers = desc_chain
+        let chain = if desc_chain.is_indirect() {
+            desc_chain
+                .new_from_indirect()
+                .map_err(|err| Error::ConvertIndirectDescriptor(err))?
```
Comment (rbradford) [inline vhost_user_fs/src/descriptor_utils.rs]: Clippy doesn't like this. Change to: `.map_err(Error::ConvertIndirectDescriptor)?`

File: vhost_user_fs/src/descriptor_utils.rs [hunk 7]
```diff
@@ -343,7 +353,14 @@ impl<'a> Writer<'a> {
     /// Construct a new Writer wrapper over `desc_chain`.
     pub fn new(mem: &'a GuestMemoryMmap, desc_chain: DescriptorChain<'a>) -> Result<Writer<'a>> {
         let mut total_len: usize = 0;
-        let buffers = desc_chain
+        let chain = if desc_chain.is_indirect() {
+            desc_chain
+                .new_from_indirect()
+                .map_err(|err| Error::ConvertIndirectDescriptor(err))?
```
Comment (rbradford) [inline vhost_user_fs/src/descriptor_utils.rs]: Ditto.

File: src/bin/vhost_user_fs.rs (hunk 3 above)
Comment (slp) [reply src/bin/vhost_user_fs.rs]: Fixed, thanks. And, while I was there, I also removed the unwraps from Reader/Writer.

File: vhost_user_fs/src/descriptor_utils.rs (hunk 6 above)
Comment (slp) [reply vhost_user_fs/src/descriptor_utils.rs]: Fixed, thanks!

**Pull request merged**